src/runtime/security/client_negotiation.cpp

```diff
@@ -36,18 +40,87 @@ void client_negotiation::start()
     list_mechanisms();
 }
 
-void client_negotiation::handle_response(error_code err, const negotiation_response &&response)
-{
-    // TBD(zlw)
-}
-
 void client_negotiation::list_mechanisms()
 {
     auto request = dsn::make_unique<negotiation_request>();
     _status = request->status = negotiation_status::type::SASL_LIST_MECHANISMS;
     send(std::move(request));
 }
 
+void client_negotiation::handle_response(error_code err, const negotiation_response &&response)
+{
+    if (err != ERR_OK) {
+        fail_negotiation();
+        return;
+    }
+
+    // make the negotiation succeed if server doesn't enable auth and the auth is not mandantory
+    if (negotiation_status::type::SASL_AUTH_DISABLE == response.status && !FLAGS_mandatory_auth) {
+        ddebug_f("{}: treat negotiation succeed as server doesn't enable it", _name);
+        succ_negotiation();
+        return;
+    }
+
+    switch (_status) {
+    case negotiation_status::type::SASL_LIST_MECHANISMS:
+        recv_mechanisms(response);
+        break;
+    case negotiation_status::type::SASL_SELECT_MECHANISMS:
+        // TBD(zlw)
+        break;
+    case negotiation_status::type::SASL_INITIATE:
+    case negotiation_status::type::SASL_CHALLENGE_RESP:
+        // TBD(zlw)
+        break;
+    default:
+        fail_negotiation();
+    }
+}
+
+void client_negotiation::recv_mechanisms(const negotiation_response &resp)
+{
+    if (resp.status != negotiation_status::type::SASL_LIST_MECHANISMS_RESP) {
+        ddebug_f("{}: get message({}) while expect({})",
+                 _name,
+                 enum_to_string(resp.status),
+                 enum_to_string(negotiation_status::type::SASL_LIST_MECHANISMS_RESP));
+        fail_negotiation();
+        return;
+    }
+
+    std::string match_mechanism = "";
```

Comment:
```suggestion
    std::string match_mechanism;
```

Reply:
I think define a variable with initial value is ok

Reply:
Default value of std::string is "", not need to assign it explicitly